From e256c9f0eb73c18921957c59e1f2f2279d561c2e Mon Sep 17 00:00:00 2001
From: Matthew Fatheree <Matthew.Fatheree@belkin.com>
Date: Tue, 13 May 2014 22:25:52 +0700
Subject: [PATCH 09/31] use /etc/preint as /sbin/init

point /sbin/init to /etc/preinit, Mamba support dual-boot feature, the
secondary firmware rootfs has the kernel argument init point to
/sbin/init while OpenWRT has init point to /etc/preinit. We link the
/sbin/init to /etc/preinit to unify the kernel init argument. The
original /sbin/init (busybox) will be call later by overriding the
pi_init_cmd variable.

Signed-off-by: Matthew Fatheree <Matthew.Fatheree@belkin.com>
---
 .../base-files/lib/preinit/01_platform_env         |    2 ++
 target/linux/armadaxp/image/Makefile               |   12 ++++++++++++
 2 files changed, 14 insertions(+)
 create mode 100644 target/linux/armadaxp/base-files/lib/preinit/01_platform_env

diff --git a/target/linux/armadaxp/base-files/lib/preinit/01_platform_env b/target/linux/armadaxp/base-files/lib/preinit/01_platform_env
new file mode 100644
index 0000000..7f6bbd7
--- /dev/null
+++ b/target/linux/armadaxp/base-files/lib/preinit/01_platform_env
@@ -0,0 +1,2 @@
+# Platform specific enviornment
+pi_init_cmd="/sbin/busybox_init init"
diff --git a/target/linux/armadaxp/image/Makefile b/target/linux/armadaxp/image/Makefile
index 7f364ff..8c1a5f1 100644
--- a/target/linux/armadaxp/image/Makefile
+++ b/target/linux/armadaxp/image/Makefile
@@ -15,6 +15,18 @@ define add_jffs2_mark
 	echo "do nothing"
 endef
 
+# override Image/mkfs/prepare/default
+define Image/mkfs/prepare/default
+	# Use symbolic permissions to avoid clobbering SUID/SGID/sticky bits
+	- $(FIND) $(TARGET_DIR) -type f -not -perm +0100 -not -name 'ssh_host*' -not -name 'shadow' -print0 | $(XARGS) -0 chmod u+rw,g+r,o+r
+	- $(FIND) $(TARGET_DIR) -type f -perm +0100 -print0 | $(XARGS) -0 chmod u+rwx,g+rx,o+rx
+	- $(FIND) $(TARGET_DIR) -type d -print0 | $(XARGS) -0 chmod u+rwx,g+rx,o+rx
+	$(INSTALL_DIR) $(TARGET_DIR)/tmp
+	chmod 1777 $(TARGET_DIR)/tmp
+	ln -sf /etc/preinit $(TARGET_DIR)/sbin/init
+	ln -sf /bin/busybox $(TARGET_DIR)/sbin/busybox_init
+endef
+
 define Image/Prepare
 	cp $(LINUX_DIR)/arch/arm/boot/uImage $(KDIR)/uImage
 	mkdir -p $(TARGET_DIR)/boot
-- 
1.7.9.5

